### ISSUE TITLE
Closing AdomdConnection throws a 'operation was canceled' error

### ISSUE THREAD

**Issue by introvert1** [bdebaere/Unofficial.Microsoft.AnalysisServices.AdomdClientNetCore#12]
When a AdomdConnection is closed, an exception is thrown. 
Exception: "System.OperationCanceledException: 'The operation was canceled.'"

This happens both when calling the close method of the AdomdConnection class and when getting out of a using(AdomdConnection) block.

**Comment by bartha-adam** [comment]
Experiencing same issue with the following code:
using (var connection = new AdomdConnection(connectionString))
{
  connection.Open();
}
Exception: "System.OperationCanceledException: 'The operation was canceled.'"

The same happens when calling connection.Close() right after connection.Open().

**Comment by bdebaere** [comment]
This will take some time to investigate.

An intermediate workaround is to not wrap your connection in a using block.

```
var connection = new AdomdConnection(connectionString);
{
    connection.Open();
}
```

**Comment by bartha-adam** [comment]
@bdebaere yes, that's what we are using + not calling the .close() method.
As a result, we're ending up with stale connections and sessions on SSAS server (around 300+ sessions and 300+ connection in dev environment).

**Comment by bdebaere** [comment]
@bartha-adam If you close the connection manually, do you still get the error? Can you paste the stack trace please?

**Comment by bartha-adam** [comment]
@bdebaere sure:
Exception message: "The operation was canceled."
Exception stacktrace:
"   at System.Net.HttpWebRequest.GetResponse()\r\n   at Microsoft.AnalysisServices.AdomdClient.HttpStream.GetResponseStream()\r\n   at Microsoft.AnalysisServices.AdomdClient.HttpStream.GetResponseDataType()\r\n   at Microsoft.AnalysisServices.AdomdClient.CompressedStream.GetResponseDataType()\r\n   at Microsoft.AnalysisServices.AdomdClient.XmlaClient.EndRequest(Boolean useBinaryXml)\r\n   at Microsoft.AnalysisServices.AdomdClient.XmlaClient.SendExecuteAndReadResponse(Boolean skipResults, Boolean throwIfError, Boolean useBinaryXml)\r\n   at Microsoft.AnalysisServices.AdomdClient.XmlaClient.EndSession(ListDictionary properties)\r\n   at Microsoft.AnalysisServices.AdomdClient.XmlaClient.Disconnect(Boolean endSession)\r\n   at Microsoft.AnalysisServices.AdomdClient.AdomdConnection.XmlaClientProvider.Microsoft.AnalysisServices.AdomdClient.AdomdConnection.IXmlaClientProviderEx.Disconnect(Boolean endSession)\r\n   at Microsoft.AnalysisServices.AdomdClient.AdomdConnection.Close(Boolean endSession)\r\n   at PortalAPI.DAL.SSASClient.Execute(MDXQuery mdxQuery) in <our code>

**Comment by bartha-adam** [comment]
@bdebaere yes, closing it manually leads to an exception. The above callstack is with calling close manually.

**Comment by bartha-adam** [comment]
@bdebaere is there a reason behind setting the timeout to 100ms when closing the connection?
Code from XmlaClient.cs:
```
private void ReadSettingsFromConfiguration()
		{
			int num = 100;
			//if (this.TryGetConfigValue<int>("AS_HttpStreamBufferSize", out num, (int val) => val >= 0 && val <= 10240))
			//{
				XmlaClient.HttpStreamBufferSize = num * 1024;
			//}
			//if (this.TryGetConfigValue<int>("AS_TcpStreamBufferSize", out num, (int val) => val >= 0 && val <= 10240))
			//{
				XmlaClient.BufferSizeForNetworkStream = num * 1024;
			//}
			//if (this.TryGetConfigValue<int>("AS_EndSessionTimeout", out num, (int val) => val >= 0 || val == -1))
			//{
				XmlaClient.EndSessionTimeout = num;
			//}
		}
```
Commented line 2227 from XmlaClient.cs (_XmlaClient.EndSessionTimeout = num;_), and Close() call does not throw anymore the System.OperationCanceledException.

**Comment by mike-beall-phs** [comment]
I am seeing the same exception happening frequently and it has destabilized our application as compared with the framework / MS AdomdClient version. 

Exception message: "The operation was canceled."
Exception stacktrace:
" at System.Net.HttpWebRequest.GetResponse()\r\n at Microsoft.AnalysisServices.AdomdClient.HttpStream.GetResponseStream()\r\n at ....

I am going to try this change and see if it fixes my issue. Will report back.

**Comment by bartha-adam** [comment]
> I am seeing the same exception happening frequently and it has destabilized our application as compared with the framework / MS AdomdClient version.
> 
> Exception message: "The operation was canceled."
> Exception stacktrace:
> " at System.Net.HttpWebRequest.GetResponse()\r\n at Microsoft.AnalysisServices.AdomdClient.HttpStream.GetResponseStream()\r\n at ....
> 
> I am going to try this change and see if it fixes my issue. Will report back.

Any update?

**Comment by mike-beall-phs** [comment]
>> Commented line 2227 from XmlaClient.cs (XmlaClient.EndSessionTimeout = num;), and Close() call does not throw anymore the System.OperationCanceledException.

I did this and rebuilt the nuget pkg and everything works correctly. The default for XmlaClient.EndSessionTimeout is 30000.

@bdebaere, any comments?

**Comment by dancundy** [comment]
I am also seeing this exception in seemingly random frequency . Closing the connection manually or automatically throws the exception. I'm having to do something horrid like below to handle the instability.

```
                    try
                    {
                        conn.Close();
                    }
                    catch (Exception e)
                    {
                        if (e.Message == "The operation was canceled.")
                        {

                        }
                        else throw;
                    }

```

**Comment by dancundy** [comment]
One thing I did think was strange is whenever I look at the state of the connection before I close, it is always already closed.

**Comment by mike-beall-phs** [comment]
@dancundy why don't you check the state before trying to close like:

                  if (connection.State != ConnectionState.Closed)
                        connection.Close();
I don't explicitly close the connections, I use this pattern:

```
            using (var connection = new AdomdConnection(connStr))
            {
                try
                {
                    connection.Open();
                   <more code .. catch>
            }  // end using
```
Using the "using" pattern, dispose is called at the end block and dispose closes the connection gracefully. The library has been very stable for us in production after I made the above fix I commented on.

Actually, I just re-read my comment and it fixes your issue:

> Commented line 2227 from XmlaClient.cs (XmlaClient.EndSessionTimeout = num;), and Close() call does not throw anymore the System.OperationCanceledException.

Mike

**Comment by mikebeall** [comment]
@dancundy why don't you check the state before trying to close like:

                  if (connection.State != ConnectionState.Closed)
                        connection.Close();
I don't explicitly close the connections, I use this pattern:

```
            using (var connection = new AdomdConnection(connStr))
            {
                try
                {
                    connection.Open();
                   <more code .. catch>
            }  // end using
```
Using the "using" pattern, dispose is called at the end block and dispose closes the connection gracefully. The library has been very stable for us in production after I made the above fix I commented on.

Actually, I just re-read my comment and it fixes your issue:

> Commented line 2227 from XmlaClient.cs (XmlaClient.EndSessionTimeout = num;), and Close() call does not throw anymore the System.OperationCanceledException.

Mike
